near-sdk/src/promise.rs

```diff
@@ -632,6 +680,54 @@ impl<T: schemars::JsonSchema> schemars::JsonSchema for PromiseOrValue<T> {
     }
 }
 
+/// A list of promises that are executed concurrently.
+///
+/// This is the return type of [`Promise::then_concurrent`] and it wraps the
+/// contained promises for a more convenient syntax when chaining calls.
+///
+/// Use [`ConcurrentPromises::join`] to create a new promises that waits for all
+/// concurrent promises and takes all their return values as inputs.
+///
+/// Use [`ConcurrentPromises::split_off`] to divide the list of promises into
+/// subgroups that can be joined independently.
+///
+/// You can also use [`ConcurrentPromises::take`] to retrieve the list of
+/// promises as a vector.
+pub struct ConcurrentPromises {
+    promises: Vec<Promise>,
+}
+
+impl ConcurrentPromises {
+    /// Create a new promises that waits for all contained concurrent promises.
+    ///
+    /// The returned promise is a [`Promise::and`] combination of all contained
+    /// promises. Chain it with a [`Promise::then`] to wait for them to finish
+    /// and receive all their outputs as inputs in a following function call.
+    pub fn join(self) -> Promise {
+        self.promises
+            .into_iter()
+            .reduce(|left, right| left.and(right))
+            .expect("cannot join empty concurrent promises")
+    }
+
+    /// Splits the contained list of promises into two at the given index,
+    /// returning a new [`ConcurrentPromises`] containing the elements in the
+    /// range [at, len).
+    ///
+    /// After the call, the original [`ConcurrentPromises`] will be left
+    /// containing the elements [0, at).
+    pub fn split_off(&mut self, at: usize) -> ConcurrentPromises {
+        let right_side = self.promises.split_off(at);
+        ConcurrentPromises { promises: right_side }
+    }
+
+    /// Take the contained concurrent promises out of the convenience wrapper
+    /// [`ConcurrentPromises`].
+    pub fn take(self) -> Vec<Promise> {
+        self.promises
+    }
```

Comment:
This might be a footgun, so I am not sure it is a good idea to expose this method. I intuitively feel that there is some corner case, but I cannot come up with one at the moment.

Reply:
Removed it from the PR, unlikely anyone will ever miss it.